File: src/main/scala/cn/academy/medicine/BuffData.scala
package cn.academy.medicine

import java.util
import java.util.{Collections, Comparator}

import cn.academy.ability.api.data.CPData
import cn.academy.core.Resources
import cn.academy.medicine.BuffData.BuffApplyData
import cn.lambdalib.annoreg.core.Registrant
import cn.lambdalib.annoreg.mc.RegPostInitCallback
import cn.lambdalib.util.datapart.{DataPart, EntityData, RegDataPart}
import cn.lambdalib.util.generic.MathUtils
import cn.lambdalib.util.helper.TickScheduler
import cpw.mods.fml.common.network.ByteBufUtils
import cpw.mods.fml.relauncher.Side
import io.netty.buffer.ByteBuf
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.nbt.{NBTTagCompound, NBTTagList}
import net.minecraft.potion.{Potion, PotionEffect}
import net.minecraft.util.{DamageSource, ResourceLocation}

import scala.collection.mutable.ArrayBuffer
import scala.reflect.ClassTag

trait Buff {

  def onBegin(player: EntityPlayer) = {}

  def onTick(player: EntityPlayer, applyData: BuffApplyData) = {}

  def onEnd(player: EntityPlayer) = {}


  def load(tag: NBTTagCompound) = {}

  def store(tag: NBTTagCompound) = {}

  val id: String

  val icon: ResourceLocation = Resources.getTexture("buff/" + id)

  val shouldDisplay: Boolean = true

}

@Registrant
@RegBuff
class BuffMedSens() extends Buff {
  private var percentage: Float = 0.0f

  override def onTick(player: EntityPlayer, applyData: BuffApplyData) = {
    percentage = math.max(0.0f, percentage - 0.0005f)
    if (percentage == 0) {
      applyData.setEnd()
    }
  }

  def increase(player: EntityPlayer, pct: Float): Unit = {
    val prev = percentage
    percentage = MathUtils.clampf(0, 1, pct + percentage)

    def trigger(test: Float) = prev < test && percentage >= test
    def damage(dmg: Float) = {
      val realdmg = math.min(player.getHealth - 1, dmg)
      player.attackEntityFrom(DamageSource.causePlayerDamage(player), realdmg)
    }
    def hunger(secs: Int) = {
      val ticks = secs * 20
      val effect = new PotionEffect(Potion.hunger.id, ticks, 1)
      player.addPotionEffect(effect)
    }

    if (trigger(1f)) {
      player.hurtResistantTime = -1
      player.setHealth(0f)
    } else if (trigger(0.8f)) {
      val cpData = CPData.get(player)

      damage(10)
      cpData.perform(cpData.getMaxOverload * 0.6f, 0)
    } else if (trigger(0.6f)) {
      damage(5)
      hunger(15)
    } else if (trigger(0.4f)) {
      damage(3)
      hunger(3)
    }
  }

  override val shouldDisplay = false
  override val id: String = "med_sensitive"
}

@Registrant
object BuffRegistry {

  private val buffTypes = new util.ArrayList[Class[_ <: Buff]]()

  def _register(klass: Class[_ <: Buff]) = buffTypes.add(klass)

  @RegPostInitCallback
  def _init() = {
    Collections.sort(buffTypes.asInstanceOf[util.List[Any]], new Comparator[Any] {
      override def compare(o1: Any, o2: Any): Int = {
        o1.toString.compareTo(o2.toString)
      }
    })
  }

  def writeBuff(buff: Buff, tag: NBTTagCompound): Unit = {
    val id = buffTypes.indexOf(buff.getClass)
    tag.setInteger("id", id)
    buff.store(tag)
  }

  def readBuff(tag: NBTTagCompound): Buff = {
    val id = tag.getInteger("id")
    val buff = buffTypes.get(id).newInstance()
    buff.load(tag)

    buff
  }

}

object BuffData {

  case class BuffApplyData(var tickLeft: Int, private var maxTicks_ : Int) {
    def isInfinite = maxTicks_ == -1

    def setEnd() = {
      if (isInfinite) {
        maxTicks_ = 10
      }
      tickLeft = 0
    }

    def maxTicks = maxTicks_

  }

  case class BuffRuntimeData(buff: Buff, applyData: BuffApplyData)

  def apply(player: EntityPlayer): BuffData = EntityData.get(player).getPart(classOf[BuffData])

}

@Registrant
@RegDataPart(value=classOf[EntityPlayer])
class BuffData extends DataPart[EntityPlayer] {
  import BuffData._
  import scala.collection.JavaConversions._

  private case class ClientFakeBuff(id: String) extends Buff {}

  setClearOnDeath()
  setClientNeedSync()
  setNBTStorage()
  setTick(true)

  private val activeBuffs = new util.ArrayList[BuffRuntimeData]()

  private val scheduler = new TickScheduler

  scheduler.every(10).atOnly(Side.SERVER).run(() => sync())

  scheduler.everyTick().atOnly(Side.SERVER).run(() => {
    val player = getEntity
    val iter = activeBuffs.iterator()
    while (iter.hasNext) {
      iter.next match {
        case BuffRuntimeData(buff, applyData) =>
          if (applyData.tickLeft > 0 || applyData.isInfinite) {
            if (!applyData.isInfinite) {
              applyData.tickLeft -= 1
            }

            buff.onTick(player, applyData)
          } else {
            buff.onEnd(player)
            iter.remove()
          }
      }
    }
  })

  def addBuffInfinite(buff: Buff) = addBuff(buff, -1)

  def addBuff(buff: Buff, maxTicks: Int) = {
    checkSide(Side.SERVER)
    activeBuffs.add(BuffRuntimeData(buff, BuffApplyData(maxTicks, maxTicks)))

    sync()
  }

  def findBuff[T <: Buff](implicit tag: ClassTag[T]): Option[T] = {
    activeBuffs.find(tag.runtimeClass.isInstance).map(_.asInstanceOf[T])
  }

  override def tick(): Unit = {
    scheduler.runTick()
  }

  override def fromByteBuf(buf: ByteBuf): Unit = {
    checkSide(Side.CLIENT)

    activeBuffs.clear()

    val count = buf.readInt
    (0 until count).foreach(_ => {
      val id = ByteBufUtils.readUTF8String(buf)
      val tickLeft = buf.readInt
      val maxTicks = buf.readInt

      activeBuffs.add(BuffRuntimeData(ClientFakeBuff(id), BuffApplyData(tickLeft, maxTicks)))
    })
  }

  override def toByteBuf(buf: ByteBuf): Unit = {
    checkSide(Side.SERVER)

    buf.writeInt(activeBuffs.size)
    activeBuffs.foreach { case BuffRuntimeData(buff, BuffApplyData(tickLeft, maxTicks)) =>
        ByteBufUtils.writeUTF8String(buf, buff.id)
        buf.writeInt(tickLeft).writeInt(maxTicks)
    }
  }

  override def toNBT(tag: NBTTagCompound) = {
    val list = new NBTTagList
    activeBuffs.foreach { case BuffRuntimeData(buff, applyData) =>
        val tag = new NBTTagCompound
        val buffTag = new NBTTagCompound
        BuffRegistry.writeBuff(buff, buffTag)
        tag.setInteger("maxTicks", applyData.maxTicks)
        tag.setInteger("tickLeft", applyData.tickLeft)
        tag.setTag("buff", buffTag)

        list.appendTag(tag)
    }
  }

  override def fromNBT(tag: NBTTagCompound) = {
    tag.getTag("buff") match {
      case list: NBTTagList =>
        (0 until list.tagCount).map(list.getCompoundTagAt)
          .foreach(tag => {
            val buffTag = tag.getTag("buff").asInstanceOf[NBTTagCompound]
            val buff = BuffRegistry.readBuff(buffTag)
            val applyData = BuffApplyData(tag.getInteger("tickLeft"), tag.getInteger("maxTicks"))

            activeBuffs.add(BuffRuntimeData(buff, applyData))
          })
      case _ =>
    }
  }

}

File: src/main/java/cn/academy/medicine/api/BuffData.java
package cn.academy.medicine.api;

import cn.lambdalib.annoreg.core.Registrant;
import cn.lambdalib.util.datapart.DataPart;
import cn.lambdalib.util.datapart.EntityData;
import cn.lambdalib.util.datapart.RegDataPart;
import cn.lambdalib.util.helper.TickScheduler;
import cpw.mods.fml.common.network.ByteBufUtils;
import cpw.mods.fml.relauncher.Side;
import io.netty.buffer.ByteBuf;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.nbt.NBTBase;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;
import scala.reflect.ClassTag;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;


@Registrant
@RegDataPart(value=EntityPlayer.class)
public class BuffData extends DataPart<EntityPlayer> {
    class ClientFakeBuff extends Buff {
        @Override
        public void onBegin(EntityPlayer player) {

        }

        @Override
        public void onTick(EntityPlayer player, BuffApplyData applyData) {

        }

        @Override
        public void onEnd(EntityPlayer player) {

        }

        @Override
        public void load(NBTTagCompound tag) {

        }

        @Override
        public void store(NBTTagCompound tag) {

        }

        public ClientFakeBuff(String id) {
            super(id);
        }
    }
    public class BuffRuntimeData {
        public Buff buff;
        public BuffApplyData applyData;
        public BuffRuntimeData(Buff buff, BuffApplyData data)
        {
            this.buff=buff;
            this.applyData=data;
        }
    }


    private List<BuffRuntimeData> activeBuffs = new ArrayList<>();

    private TickScheduler scheduler = new TickScheduler();

    public BuffData(){
        setClearOnDeath();
        setClientNeedSync();
        setNBTStorage();
        setTick(true);
        scheduler.every(10).atOnly(Side.SERVER).run(this::sync);

        scheduler.everyTick().atOnly(Side.SERVER).run(() -> {
                EntityPlayer player = getEntity();
                Iterator iter = activeBuffs.iterator();
                while (iter.hasNext()) {
                    if(iter.next() instanceof BuffRuntimeData)
                    {
                        BuffRuntimeData data = (BuffRuntimeData)iter.next();
                        if (data.applyData.tickLeft > 0 || data.applyData.isInfinite()) {
                            if (!data.applyData.isInfinite()) {
                                data.applyData.tickLeft -= 1;
                            }
                            data.buff.onTick(player, data.applyData);
                        } else {
                            data.buff.onEnd(player);
                            iter.remove();
                        }
                    }
                }
            });
    }


    public void addBuffInfinite(Buff buff){
        addBuff(buff, -1);
    }

    public void addBuff(Buff buff, int maxTicks){
        checkSide(Side.SERVER);
        activeBuffs.add(new BuffRuntimeData(buff, new BuffApplyData(maxTicks, maxTicks)));

        sync();
    }

    public <T extends Buff> T findBuff(ClassTag<T> tag){
        for(BuffRuntimeData buff : activeBuffs)
        {
            if(tag.runtimeClass().isInstance(buff)){
                return (T)(buff.buff);
            }
        }
        return null;
    }

    public void tick(){
        scheduler.runTick();
    }

    public void fromByteBuf(ByteBuf buf){
        checkSide(Side.CLIENT);
        activeBuffs.clear();

        int count = buf.readInt();
        for(int i=0;i<count;i++){
            String id = ByteBufUtils.readUTF8String(buf);
            int tickLeft = buf.readInt();
            int maxTicks = buf.readInt();
            activeBuffs.add(new BuffRuntimeData(new ClientFakeBuff(id), new BuffApplyData(tickLeft, maxTicks)));
        }
    }

    public void toByteBuf(ByteBuf buf){
        checkSide(Side.SERVER);
        buf.writeInt(activeBuffs.size());
        for(BuffRuntimeData data:activeBuffs)
        {
            ByteBufUtils.writeUTF8String(buf, data.buff.id);
            buf.writeInt(data.applyData.tickLeft).writeInt(data.applyData.maxTicks());
        }
    }

    public void toNBT(NBTTagCompound tag){
        NBTTagList list = new NBTTagList();
        for(BuffRuntimeData data:activeBuffs)
        {
            NBTTagCompound tag_ = new NBTTagCompound();
            NBTTagCompound buffTag = new NBTTagCompound();
            BuffRegistry.writeBuff(data.buff, buffTag);
            tag_.setInteger("maxTicks", data.applyData.maxTicks());
            tag_.setInteger("tickLeft", data.applyData.tickLeft);
            tag_.setTag("buff", buffTag);

            list.appendTag(tag_);
        }
        tag.setTag("buff",list);
    }

    public void  fromNBT(NBTTagCompound tag){
        NBTBase tBase = tag.getTag("buff");
        if (tBase instanceof NBTTagList){
            for(int i=0;i<((NBTTagList) tBase).tagCount();i++)
            {
                NBTTagCompound tag_=((NBTTagList) tBase).getCompoundTagAt(i);
                NBTTagCompound buffTag = (NBTTagCompound) tag.getTag("buff");
                Buff buff = BuffRegistry.readBuff(buffTag);
                BuffApplyData applyData = new BuffApplyData(tag.getInteger("tickLeft"),
                        tag.getInteger("maxTicks"));
                activeBuffs.add(new BuffRuntimeData(buff, applyData));

            }
        }
    }

    public static BuffData apply(EntityPlayer player)
    {
        return EntityData.get(player).getPart(BuffData.class);
    }

}
